src/clang_parser.cpp

```diff
@@ -286,55 +292,80 @@ void ClangParser::parse(ast::Program *program, BPFtrace &bpftrace)
+  do {
+    clang_visitChildren(
+        cursor,
+        [](CXCursor c, CXCursor parent, CXClientData client_data)
         {
-          std::string macro_name;
-          std::string macro_value;
-          if (translateMacro(c, macro_name, macro_value)) {
-            auto &macros = static_cast<BPFtrace*>(client_data)->macros_;
-            macros[macro_name] = macro_value;
+          if (clang_getCursorKind(c) == CXCursor_MacroDefinition)
+          {
+            std::string macro_name;
+            std::string macro_value;
+            if (translateMacro(c, macro_name, macro_value)) {
+              auto &macros = static_cast<BPFtrace*>(client_data)->macros_;
+              macros[macro_name] = macro_value;
+            }
+            return CXChildVisit_Recurse;
           }
-          return CXChildVisit_Recurse;
-        }
-
-        if (clang_getCursorKind(parent) != CXCursor_StructDecl &&
-            clang_getCursorKind(parent) != CXCursor_UnionDecl)
-          return CXChildVisit_Recurse;
 
-        if (clang_getCursorKind(c) == CXCursor_FieldDecl)
-        {
-          auto &structs = static_cast<BPFtrace*>(client_data)->structs_;
-          auto struct_name = get_parent_struct_name(c);
-          auto ident = get_clang_string(clang_getCursorSpelling(c));
-          auto offset = clang_Cursor_getOffsetOfField(c) / 8;
-          auto type = clang_getCanonicalType(clang_getCursorType(c));
+          if (clang_getCursorKind(parent) != CXCursor_StructDecl &&
+              clang_getCursorKind(parent) != CXCursor_UnionDecl)
+            return CXChildVisit_Recurse;
 
           auto ptype = clang_getCanonicalType(clang_getCursorType(parent));
           auto ptypestr = get_clang_string(clang_getTypeSpelling(ptype));
           auto ptypesize = clang_Type_getSizeOf(ptype);
 
-          if(clang_Cursor_isAnonymous(parent))
-            offset = get_indirect_field_offset(c);
+          if (clang_getCursorKind(c) == CXCursor_StructDecl ||
+              clang_getCursorKind(c) == CXCursor_UnionDecl) {
+            auto struct_name = get_clang_string(clang_getTypeSpelling(clang_getCanonicalType(clang_getCursorType(c))));
+            indirect_structs[struct_name] = c;
+            unvisited_indirect_structs.insert(struct_name);
 
-          if (struct_name == "")
-            struct_name = ptypestr;
-          remove_struct_prefix(struct_name);
+            return CXChildVisit_Continue;
+          }
+
+          if (clang_getCursorKind(c) == CXCursor_FieldDecl)
+          {
+            auto &structs = static_cast<BPFtrace*>(client_data)->structs_;
+            auto struct_name = get_parent_struct_name(c);
+            auto ident = get_clang_string(clang_getCursorSpelling(c));
+            auto offset = clang_Cursor_getOffsetOfField(c) / 8;
+            auto type = clang_getCanonicalType(clang_getCursorType(c));
+            auto typestr = get_clang_string(clang_getTypeSpelling(type));
+
+            if (indirect_structs.count(typestr))
```

Comment:
When would we hit this? Could you add a comment as it's not immediately clear to me

Reply:
Following the example above, would be in this case:

```
struct Foo {
  int field1;
  struct {
    int field2;
  } <field3>;
  char field4;
}
```

We saved the possibly indirect struct in our previous step, and then when we moved the cursor we realized that we were actually declaring a field of the parent struct. In this case we don't want to treat the struct as indirect, so we remove it from our list of indirect structs.

Looking at this again, this seems to be incomplete, as we'll never parse the unnamed struct in that example (unless I'm missing something). I'll investigate.

Reply:
Apparently it works, but I don't know why :upside_down_face: